src/gt4py/next/iterator/transforms/infer_domain.py

```diff
@@ -0,0 +1,161 @@
+                original_domains[key] = domain_union([original_domains[key], value])
+            else:
+                original_domains[key] = value
+
+        return {
+            key: domain_union(value) if isinstance(value, list) else value
+            for key, value in original_domains.items()
+        }
+
+    @classmethod
+    def _translate_domain(
+        cls,
+        symbolic_domain: SymbolicDomain,
+        shift: Tuple[itir.OffsetLiteral, int],
+        dims: List[str],
+        offset_provider: Dict[str, Dimension],
+    ) -> SymbolicDomain:
+        new_ranges = {dim: symbolic_domain.ranges[dim] for dim in dims}
+        if shift:
+            off, val = shift
+            current_dim = offset_provider[off.value]
+            new_ranges[current_dim.value] = SymbolicRange.translate(
+                symbolic_domain.ranges[current_dim.value], val.value
+            )
+        return SymbolicDomain("cartesian_domain", new_ranges)
+
+    @classmethod
+    def infer_as_fieldop(
+        cls,
+        applied_fieldop: itir.FunCall,
+        input_domain: SymbolicDomain,
+        offset_provider: Dict[str, Dimension],
+    ) -> Tuple[itir.FunCall, Dict[str, SymbolicDomain]]:  # todo: test scan operator
+        assert isinstance(applied_fieldop, itir.FunCall) and isinstance(
+            applied_fieldop.fun, itir.FunCall
+        )
+        assert applied_fieldop.fun.fun == im.ref("as_fieldop")
+
+        stencil, inputs = applied_fieldop.fun.args[0], applied_fieldop.args
+
+        inputs_node = []
+        accessed_domains = {}
+
+        # Set inputs for StencilClosure node by replacing FunCalls with temporary SymRefs
+        tmp_counter = 0
+        for in_field in inputs:
+            if isinstance(in_field, itir.FunCall):
+                in_field.id = im.ref(f"__dom_inf_{tmp_counter}")
+                inputs_node.append(im.ref(in_field.id))
+                accessed_domains[str(in_field.id)] = []
+                tmp_counter += 1
+            else:
+                inputs_node.append(in_field)
+                accessed_domains[str(in_field.id)] = []
+
+        out_field_name = "tmp"  # todo: can this be derived from somewhere?
+
+        symbolic_domain = cls._get_symbolic_domain(input_domain)
+
+        # TODO: until TraceShifts directly supporty stencils we just wrap our expression into a dummy closure in this helper function.
+        def trace_shifts(stencil: itir.Expr, inputs: list[itir.Expr], domain: itir.Expr):
+            node = itir.StencilClosure(
+                stencil=stencil,
+                inputs=inputs,
+                output=im.ref(out_field_name),
+                domain=domain,
+            )
+            return TraceShifts.apply(node)
+
+        # Extract the shifts and translate the domains accordingly
+        shifts_results = trace_shifts(stencil, inputs_node, SymbolicDomain.as_expr(symbolic_domain))
+        dims = cls._extract_axis_dims(SymbolicDomain.as_expr(symbolic_domain))
```

Comment:
Should be possible to remove the function and also absorb it into translate domain.